pkg/backend/net/net.go

```diff
@@ -237,7 +237,10 @@ func (b Backend) do(ctx context.Context, trace types.Trace, req *http.Request) (
 
 	go func() {
 		resp, err := b.client.Do(req)
-		ch <- requestRes{resp: resp, err: err}
+		select {
+		case ch <- requestRes{resp: resp, err: err}:
```

Comment:
Does this select then change the behavior of the one below in a predictable way? IIRC the runtime doesn't make guarantees about which channel it reads from if more than one are ready (double-IIRC the limiter code assumes that).

In any case, if the ctx cancels then the original request has gotten/is getting/will get an error response, so it matters less exactly what we do to the response here than that we clean up any goroutines and resources we've taken out.

Reply:
But there is a race: Given that ctx is canceled, both the reader of the `requestRes` from `ch` below and the `b.client.Do` stop doing what they are doing. That is: the reader below may return upon reading on `ctx.Done()` and will not try to read the `requestRes`. At the same time `b.client.Do` terminates with a `context.Canceled`, tries to send that into `ch` which has no reader anymore and block forever. Thence there is a goroutine leak. 

A possible regression might be that we lose an error that appears right before a timeout by coming late to the race. That is an unlikely coincidence and the lost error may not be worth reporting then?

I have the feeling we're saying the same thing are we?